Realm/Realm/DatabaseTypes/KeyPathCollection.cs

```diff
@@ -0,0 +1,114 @@
+﻿////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2023 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License")
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+
+namespace Realms;
+
+//TODO Add docs for this and the following types
+public class KeyPathsCollection : IEnumerable<KeyPath>
+{
+    private IEnumerable<KeyPath> _collection;
+
+    private static readonly KeyPathsCollection _shallow = new KeyPathsCollection(KeyPathsCollectionType.Shallow);
+    private static readonly KeyPathsCollection _default = new KeyPathsCollection(KeyPathsCollectionType.Default);
+
+    internal KeyPathsCollectionType Type { get; set; }
+
+    private KeyPathsCollection(KeyPathsCollectionType type, IEnumerable<KeyPath>? collection = null)
```

Comment:
I think taking an enumerable here is not ideal - it's not wrong, but makes the API semantics a bit awkward - i.e. we enumerate it multiple times and if it was a generator-based enumerable or something, we'd get weird results. Perhaps using `ICollection` is a better choice here and both array and list conform to the interface.

Reply:
Reasonable